pkg/serverless/trace/cold_start_span_creator.go

```diff
@@ -35,45 +33,100 @@ const (
 var functionName = os.Getenv(functionNameEnvVar)
 
 type ColdStartSpanCreator struct {
-	executionContext *executioncontext.ExecutionContext
-	traceAgent       *ServerlessTraceAgent
-	createSpan       *sync.Once
+	TraceAgent            *ServerlessTraceAgent
+	createSpan            sync.Once
+	LambdaSpanChan        <-chan *pb.Span
+	InitDurationChan      <-chan float64
+	syncSpanDurationMutex sync.Mutex
+	ColdStartSpanId       uint64
+	lambdaSpan            *pb.Span
+	initDuration          float64
+	StopChan              chan struct{}
 }
 
-func (c *ColdStartSpanCreator) create(lambdaSpan *pb.Span) {
-	// Prevent infinite loop from SpanModifier call
-	if lambdaSpan.Name == "aws.lambda.cold_start" {
+func (c *ColdStartSpanCreator) Run() {
+	go func() {
+		for {
+			select {
+			case traceAgentSpan := <-c.LambdaSpanChan:
+				c.handleLambdaSpan(traceAgentSpan)
+
+			case initDuration := <-c.InitDurationChan:
+				c.handleInitDuration(initDuration)
+
+			case <-c.StopChan:
+				log.Debugf("[ColdStartCreator] - shutting down")
+				return
+			}
+		}
+	}()
+}
+
+func (c *ColdStartSpanCreator) Stop() {
+	log.Debugf("[ColdStartCreator] - sending shutdown msg")
+	c.StopChan <- struct{}{}
+}
```

Comment:
Just FYI and I feel like sharing. No changes requested.

I know somewhere else you did

```go
if coldStartSpanCreator != nil {
    coldStartSpanCreator.Stop()
}
```

Another cool alternative is to remove that `nil` check and add it to this `Stop()` method instead.

```go
func (c *ColdStartSpanCreator) Stop() {
    if c == nil {
        return
    }
	log.Debugf("[ColdStartCreator] - sending shutdown msg")
	c.StopChan <- struct{}{}
}
```

Since the method has a pointer receiver, the `Stop()` method will always exist, even if the `ColdStartSpanCreator` itself is nil.